### PR TITLE
Setup Go Server codegen

### PR DESCRIPTION
**Pull request by syall**
# The branch `feat-svcgen` should NOT be merged into main; everything in the branch is a POC.

*Issue #, if available:*

N/A.

*Description of changes:*

Setup Go Server codegen

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/GoIntegration.java [hunk 1]
```diff
@@ -49,6 +50,10 @@ default byte getOrder() {
         return 0;
     }
 
+    default ArtifactType getArtifactType() {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/GoIntegration.java:53]: Is it possible for a plugin to work for either artifact type? What's this used for?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java [hunk 2]
```diff
@@ -70,7 +71,8 @@
  * <p>Reserved words for Go are automatically escaped so that they are
  * suffixed with "_". See "reserved-words.txt" for the list of words.
  */
-final class SymbolVisitor implements SymbolProvider, ShapeVisitor<Symbol> {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SymbolVisitor.java:73]: nit - is this really a SymbolVisitor or a ShapeVisitor?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenVisitor.java [hunk 3]
```diff
@@ -79,8 +80,10 @@ final class CodegenVisitor extends ShapeVisitor.Default<Void> {
         LOGGER.info("Attempting to discover GoIntegration from the classpath...");
         ServiceLoader.load(GoIntegration.class, loader)
                 .forEach(integration -> {
-                    LOGGER.info(() -> "Adding GoIntegration: " + integration.getClass().getName());
-                    integrations.add(integration);
+                    if (integration.getArtifactType().equals(ArtifactType.CLIENT)) {
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenVisitor.java:83]: Is there any kind of future refactoring that removes the need to add these kinds of conditionals inside of codegen?

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/GoServerCodegenPlugin.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    private static final Logger LOGGER = Logger.getLogger(GoServerCodegenPlugin.class.getName());
+
+    @Override
+    public String getName() {
+        return "go-server-codegen";
+    }
+
+    @Override
+    public void execute(PluginContext context) {
+        String onlyBuild = System.getenv("SMITHY_GO_BUILD_API");
```
Comment (mtdowling) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/GoServerCodegenPlugin.java:40]: What is going on here? Feels like this needs a comment at least.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/ServerCodegenVisitor.java [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+        protocolGenerator = resolveProtocolGenerator(integrations, model, service, settings);
+        applicationProtocol = protocolGenerator.getApplicationProtocol();
+
+        writers = new GoDelegator(fileManifest, symbolProvider);
+
+        // TODO(SSDK): do we need this?
+        protocolDocumentGenerator = new ProtocolDocumentGenerator(settings, model, writers);
+
+        // TODO(SSDK): do we need this?
+        this.eventStreamGenerator = new EventStreamGenerator(settings, model, writers, symbolProvider, service);
```
Comment (lucix-aws) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/ServerCodegenVisitor.java:167]: FWIW we probably won't get to event streams in our current scope

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/ServerProtocolGenerator.java [hunk 6]
```diff
@@ -15,9 +15,19 @@
...
 package software.amazon.smithy.go.codegen.service;
 
+import software.amazon.smithy.go.codegen.ApplicationProtocol;
 import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.SmithyInternalApi;
 
-public interface ProtocolGenerator {
+@SmithyInternalApi
+public interface ServerProtocolGenerator {
```
Comment (lucix-aws) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/service/ServerProtocolGenerator.java:24]: Was there a reason we moved this?

**Pull request merged**